### ISSUE TITLE
Build failures on 32 bit architectures (error: size of unnamed array is negative)

### ISSUE THREAD

**Issue by sebastic** [PDLPorters/pdl#308]
The Debian package build for PDL 2.040 failed on 32 bit achitectures:

 https://buildd.debian.org/status/package.php?p=pdl&suite=experimental

From the [i386 buildlog](https://buildd.debian.org/status/fetch.php?pkg=pdl&arch=armhf&ver=1%3A2.040-1%7Eexp1&stamp=1619587306&raw=0):
```make
make[5]: Entering directory '/<<PKGBUILDDIR>>/Libtmp/GSL/LINALG'
"/usr/bin/perl" "-I../../../blib/arch" "-I../../../blib/lib" "-MPDL::PP qw[PDL::GSL::LINALG PDL::GSL::LINALG LINALG ]" gsl_linalg.pd
Running Mkbootstrap for LINALG ()
chmod 644 "LINALG.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- LINALG.bs ../../../blib/arch/auto/PDL/GSL/LINALG/LINALG.bs 644
touch LINALG.xs
"/usr/bin/perl" "/usr/share/perl/5.32.1/ExtUtils/xsubpp"  -typemap '/usr/share/perl/5.32/ExtUtils/typemap' -typemap '/<<PKGBUILDDIR>>/Basic/Core/typemap.pdl'  LINALG.xs > LINALG.xsc
cp LINALG.pm ../../../blib/lib/PDL/GSL/LINALG.pm
mv LINALG.xsc LINALG.c
i686-linux-gnu-gcc -c  "-I/<<PKGBUILDDIR>>/Basic/Core"   -I/usr/include -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2   -DVERSION=\"2.040\" -DXS_VERSION=\"2.040\" -fPIC "-I/usr/lib/i386-linux-gnu/perl/5.32/CORE"   LINALG.c
../../../blib/lib/PDL/PP/PDLCode.pm: In function ‘pdl_LU_decomp_readdata’:
../../../blib/lib/PDL/PP.pm:1340:52: error: size of unnamed array is negative
 1340 |   pp_addpm {At => 'Top'}, <<EOF;
      |                                                    ^
../../../blib/lib/PDL/PP/PDLCode.pm:571:1: note: in expansion of macro ‘static_assert_linalg’
  571 |      '}',
      | ^    ~~~~           
../../../blib/lib/PDL/PP.pm:1351:10: warning: assignment to ‘size_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘PDL_Indx *’ {aka ‘long long int *’} [-Wincompatible-pointer-types]
 1351 | $SIG{__DIE__} = sub {print Carp::longmess(@_); die;}
      |          ^
../../../blib/lib/PDL/PP/PDLCode.pm:576:1: note: in expansion of macro ‘PERM_SETUP’
  576 |     }
      | ^   ~     
../../../blib/lib/PDL/PP.pm:1340:52: error: size of unnamed array is negative
 1340 |   pp_addpm {At => 'Top'}, <<EOF;
      |                                                    ^
../../../blib/lib/PDL/PP/PdlParObj.pm:421:1: note: in expansion of macro ‘static_assert_linalg’
../../../blib/lib/PDL/PP.pm:1351:10: warning: assignment to ‘size_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘PDL_Indx *’ {aka ‘long long int *’} [-Wincompatible-pointer-types]
 1351 | $SIG{__DIE__} = sub {print Carp::longmess(@_); die;}
      |          ^
../../../blib/lib/PDL/PP/PdlParObj.pm:426:1: note: in expansion of macro ‘PERM_SETUP’
../../../blib/lib/PDL/PP/PDLCode.pm: In function ‘pdl_LU_solve_readdata’:
../../../blib/lib/PDL/PP.pm:1351:10: warning: assignment to ‘size_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘PDL_Indx *’ {aka ‘long long int *’} [-Wincompatible-pointer-types]
 1351 | $SIG{__DIE__} = sub {print Carp::longmess(@_); die;}
      |          ^
../../../blib/lib/PDL/PP/PDLCode.pm:582:1: note: in expansion of macro ‘PERM_SETUP’
  582 | #
      | ^         
../../../blib/lib/PDL/PP.pm:1351:10: warning: assignment to ‘size_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘PDL_Indx *’ {aka ‘long long int *’} [-Wincompatible-pointer-types]
 1351 | $SIG{__DIE__} = sub {print Carp::longmess(@_); die;}
      |          ^
../../../blib/lib/PDL/PP/PdlParObj.pm:427:1: note: in expansion of macro ‘PERM_SETUP’
LINALG.xs: In function ‘boot_PDL__GSL__LINALG’:
LINALG.xs:2394:23: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘I32’ {aka ‘long int’} [-Wformat=]
 2394 |      Perl_croak(aTHX_ "[PDL->Version: %d PDL_CORE_VERSION: %d XS_VERSION: %s] PDL::GSL::LINALG needs to be recompiled against the newly installed PDL", PDL->Version, PDL_CORE_VERSION, XS_VERSION);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~
      |                                                                                                                                                            |
      |                                                                                                                                                            I32 {aka long int}
make[5]: *** [Makefile:325: LINALG.o] Error 1
make[5]: Leaving directory '/<<PKGBUILDDIR>>/Libtmp/GSL/LINALG'
```

**Comment by mohawk2** [comment]
Thanks for the report. It seems like this allegedly 32-bit system still has `long long int`. I need to figure out what to do about that!

**Comment by mohawk2** [comment]
I believe this is fixed in f45cc5f3e236bbdf28b3fa68dc79fafcf308ad61 and I will be making a second dev-release soon with a view to releasing a new main-release thereafter. Thank you for the continued care and keeping this project honest :-)

I'll close this but please reopen or comment if you believe problems remain.

**Comment by sebastic** [comment]
Thanks for your work.

Confirmed fixed with the changes from f45cc5f3e236bbdf28b3fa68dc79fafcf308ad61 on i386, armel, and armhf:

 https://buildd.debian.org/status/package.php?p=pdl&suite=experimental

**Comment by mohawk2** [comment]
Thanks for the feedback! 2.041 has gone out with these changes.
